### PR TITLE
gather, scatter with gpu support, passed python test

### PR DESCRIPTION
**Pull request by zchen0211**

### CODE REVIEW

File: paddle/operators/gather.cu.h [hunk 1]
```diff
@@ -0,0 +1,84 @@
...
+
+/**
+ * A thin wrapper on gpu tensor
+ * Return a new tensor from source tensor, gathered according to index
+ * input[src]: type-T source Tensor
+ * input[index]: type-int index Tensor (1-D)
+ * return: output tensor
+ */
+template <typename T>
+void GPUTGather(const Place& place, const Tensor* src, const Tensor* index,
```
Comment (QiJune) [inline paddle/operators/gather.cu.h]: GPUTGather name is not clear, and can we merge GPUGather GPUTGather together? 

And we'd better change the parameter Place of GPUTGather to parameter CUDADeviceContext. CUDADeviceContext has a CUDA stream, we should launch a CUDA kernel on specific CUDA stream.

Comment (zchen0211) [reply paddle/operators/gather.cu.h]: Great idea. Working on it now...

**Comment by QiJune** [comment]
Please merge latest develop branch first

File: paddle/operators/gather.h [hunk 2]
```diff
@@ -24,33 +24,19 @@ limitations under the License. */
...
  * Return a new tensor from source tensor, gathered according to index
  * input[src]: type-T source Tensor
  * input[index]: type-int index Tensor (1-D)
  * return: output tensor
  */
 template <typename T>
-void Gather(const platform::Place& place, const paddle::framework::Tensor* src,
-            const paddle::framework::Tensor* index,
-            paddle::framework::Tensor* output) {
+void CPUGather(const platform::Place& place,
```
Comment (QiJune) [inline paddle/operators/gather.h]: It's better unify the parameter of CPUGather and GPUGather, the first parameter should be DeviceContext
```
void CPUGather(const platform::DeviceContext& ctx...)
```

File: paddle/operators/scatter.h [hunk 3]
```diff
@@ -24,63 +24,44 @@ namespace paddle {
...
  * Return a updated tensor from source tensor, scattered according to index:
- * dst[i] += src[index[i]]
+ * dst[i] = src[index[i]]
  * input[src]: type-T source Tensor
  * input[index]: type-int index Tensor (1-D)
  * return: output tensor
  */
 template <typename T>
-void ScatterUpdate(const platform::Place& place,
+void ScatterAssign(const platform::Place& place,
```
Comment (QiJune) [inline paddle/operators/scatter.h]: The same with CPUGather

File: paddle/operators/gather.cu.h [hunk 4]
```diff
@@ -0,0 +1,79 @@
...
+
+/**
+ * A thin wrapper on gpu tensor
+ * Return a new tensor from source tensor, gathered according to index
+ * input[src]: type-T source Tensor
+ * input[index]: type-int index Tensor (1-D)
+ * return: output tensor
+ */
+template <typename T>
+void GPUGather(const platform::DeviceContext& ctx, const Tensor* src,
```
Comment (QiJune) [inline paddle/operators/gather.cu.h]: If the parameter is input, we should take `const T&`;
If the parameter is output, we should take `T*`;
If the parameter is both input and output, we should take `T*`;

Please refer to https://google.github.io/styleguide/cppguide.html#Reference_Arguments

File: paddle/operators/gather.cu.h [hunk 5]
```diff
@@ -0,0 +1,79 @@
...
+               const Tensor* index, Tensor* output) {
+  // PADDLE_ENFORCE(platform::is_gpu_place(place));
```
Comment (QiJune) [inline paddle/operators/gather.cu.h]:
```
PADDLE_ENFORCE(platform::is_gpu_place(place));
```

File: paddle/operators/scatter.cu.h [hunk 6]
```diff
@@ -0,0 +1,80 @@
...
+/**
+ * A thin wrapper on gpu tensor
+ * Return a new updated tensor from source tensor, scatter-assigned according to
+ * index
+ * input[src]: type-T source Tensor
+ * input[index]: type-int index Tensor (1-D)
+ * return: output tensor
+ */
+template <typename T>
+void GPUScatterAssign(const platform::DeviceContext& ctx,
```
Comment (QiJune) [inline paddle/operators/scatter.cu.h]: The same as CPUGather, input should be `const T&`

**Pull request merged**